File: day13/main.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	dfa := NewDFA[TKind, byte](tKindZero)
	{
		dfaNum := NewDFA[TKind, byte](tKindNum)
		dfa.AddDFA([]byte("0123456789"), dfaNum)
		dfaNum.AddDFA([]byte("0123456789"), dfaNum)

		dfa.AddPath([]byte("["), tKindOpen, tKindZero)
		dfa.AddPath([]byte("]"), tKindClose, tKindZero)
		dfa.AddPath([]byte(","), tKindComma, tKindZero)
	}
	lexer := NewLexer(dfa, tKindZero, tKindEOS, map[TKind]struct{}{
		tKindComma: {},
	})

	div1 := &Signal{isList: true, list: []*Signal{{num: 2}}}
	div2 := &Signal{isList: true, list: []*Signal{{num: 6}}}

	signals := []*Signal{div1, div2}

	var left *Signal

	count := 0
	numPairs := 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 {
			left = nil
			continue
		}
		tokens, err := lexer.Tokenize(line)
		if err != nil {
			log.Fatalln(err)
		}
		sig, rest, err := parseSignal(tokens)
		if err != nil {
			log.Fatalln(err)
		}
		if len(rest) != 1 || rest[0].Kind != tKindEOS {
			log.Fatalln("Did not parse all tokens")
		}
		if left == nil {
			left = sig
		} else {
			numPairs++
			if v, ok := compareSigs(left, sig); ok && v {
				count += numPairs
			}
		}
		signals = append(signals, sig)
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Part 1:", count)

	sort.Slice(signals, func(i, j int) bool {
		v, ok := compareSigs(signals[i], signals[j])
		return v && ok
	})
	div1Idx := sort.Search(len(signals), func(i int) bool {
		if eqSigs(div1, signals[i]) {
			return true
		}
		v, ok := compareSigs(div1, signals[i])
		return v && ok
	})
	if div1Idx >= len(signals) || !eqSigs(div1, signals[div1Idx]) {
		log.Fatalln("Divider missing")
	}
	div2Idx := sort.Search(len(signals), func(i int) bool {
		if eqSigs(div2, signals[i]) {
			return true
		}
		v, ok := compareSigs(div2, signals[i])
		return v && ok
	})
	if div2Idx >= len(signals) || !eqSigs(div2, signals[div2Idx]) {
		log.Fatalln("Divider missing")
	}
	fmt.Println("Part 2:", (div1Idx+1)*(div2Idx+1))
}

func eqSigs(left, right *Signal) bool {
	if left.isList != right.isList {
		return false
	}
	if !left.isList {
		return left.num == right.num
	}
	if len(left.list) != len(right.list) {
		return false
	}
	for n, i := range left.list {
		if !eqSigs(i, right.list[n]) {
			return false
		}
	}
	return true
}

func compareSigs(left, right *Signal) (bool, bool) {
	if !left.isList && !right.isList {
		if left.num == right.num {
			return false, false
		}
		return left.num < right.num, true
	}
	if left.isList && right.isList {
		for i := 0; i < len(left.list) && i < len(right.list); i++ {
			v, ok := compareSigs(left.list[i], right.list[i])
			if ok {
				return v, true
			}
		}
		if len(left.list) == len(right.list) {
			return false, false
		}
		return len(left.list) < len(right.list), true
	}
	if !left.isList {
		left = &Signal{
			isList: true,
			list:   []*Signal{left},
		}
	} else {
		right = &Signal{
			isList: true,
			list:   []*Signal{right},
		}
	}
	return compareSigs(left, right)
}

var (
	ErrorParse = errors.New("Failed to parse")
)

type (
	Signal struct {
		num    int
		isList bool
		list   []*Signal
	}
)

type (
	TKind int
)

const (
	tKindZero TKind = iota
	tKindEOS
	tKindNum
	tKindOpen
	tKindClose
	tKindComma
)

func parseSignal(tokens []Token[TKind, byte]) (*Signal, []Token[TKind, byte], error) {
	if len(tokens) == 0 {
		return nil, tokens, fmt.Errorf("No tokens: %w", ErrorParse)
	}
	top := tokens[0]
	switch top.Kind {
	case tKindNum:
		{
			num, err := strconv.Atoi(string(top.Val))
			if err != nil {
				return nil, tokens, fmt.Errorf("%w: %v", ErrorParse, err)
			}
			return &Signal{
				num: num,
			}, tokens[1:], nil
		}
	case tKindOpen:
		{
			tokens = tokens[1:]
			var signals []*Signal
			for {
				if len(tokens) == 0 {
					return nil, tokens, fmt.Errorf("No tokens: %w", ErrorParse)
				}
				if tokens[0].Kind == tKindClose {
					tokens = tokens[1:]
					break
				}
				sig, rest, err := parseSignal(tokens)
				if err != nil {
					return nil, rest, err
				}
				signals = append(signals, sig)
				tokens = rest
			}
			return &Signal{
				isList: true,
				list:   signals,
			}, tokens, nil
		}
	default:
		return nil, tokens, errors.New("Failed to parse")
	}
}

File: day13/src/main.rs
use std::cmp::Ordering;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;
use std::iter::Peekable;
use std::vec::IntoIter;

const PUZZLEINPUT: &str = "input.txt";

type BoxError = Box<dyn std::error::Error>;
type BoxResult<T> = Result<T, BoxError>;

fn main() -> BoxResult<()> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let div1 = Signal::List(vec![Signal::Num(2)]);
    let div2 = Signal::List(vec![Signal::Num(6)]);
    let mut signals = vec![
        Signal::List(vec![Signal::Num(2)]),
        Signal::List(vec![Signal::Num(6)]),
    ];

    let mut left = None;

    let mut count = 0;
    let mut num_pairs = 0;
    for line in reader.lines() {
        let line = line?;
        if line.len() == 0 {
            left = None;
            continue;
        }
        let tokens = tokenize(line.into_bytes().into_iter().peekable())?;
        let (sig, _) = parse_tokens(tokens.into_iter().peekable())?;
        match left {
            None => left = Some(sig),
            Some(l) => {
                num_pairs += 1;
                if let Some(v) = compare_sigs(&l, &sig) {
                    if v {
                        count += num_pairs;
                    }
                }
                left = None;
                signals.push(l);
                signals.push(sig);
            }
        }
    }

    println!("Part 1: {}", count);

    signals.sort_unstable_by(cmp_sigs);

    let div1_idx = signals
        .binary_search_by(|i| cmp_sigs(i, &div1))
        .or_else::<BoxError, _>(|_| Err("Divider missing".into()))?;

    let div2_idx = signals
        .binary_search_by(|i| cmp_sigs(i, &div2))
        .or_else::<BoxError, _>(|_| Err("Divider missing".into()))?;

    println!("Part 2: {}", (div1_idx + 1) * (div2_idx + 1));

    Ok(())
}

fn cmp_sigs(left: &Signal, right: &Signal) -> Ordering {
    if eq_sigs(left, right) {
        Ordering::Equal
    } else {
        match compare_sigs(left, right) {
            Some(v) => {
                if v {
                    Ordering::Less
                } else {
                    Ordering::Greater
                }
            }
            None => Ordering::Equal,
        }
    }
}

fn eq_sigs(left: &Signal, right: &Signal) -> bool {
    match (left, right) {
        (Signal::Num(a), Signal::Num(b)) => a == b,
        (Signal::List(a), Signal::List(b)) => {
            if a.len() != b.len() {
                return false;
            }
            for (l, r) in a.iter().zip(b.iter()) {
                if !eq_sigs(l, r) {
                    return false;
                }
            }
            true
        }
        _ => false,
    }
}

fn compare_sigs(left: &Signal, right: &Signal) -> Option<bool> {
    match (left, right) {
        (Signal::Num(a), Signal::Num(b)) => {
            if a == b {
                None
            } else {
                Some(a < b)
            }
        }
        (Signal::List(a), Signal::List(b)) => {
            for (l, r) in a.iter().zip(b.iter()) {
                if let Some(v) = compare_sigs(l, r) {
                    return Some(v);
                }
            }
            if a.len() == b.len() {
                None
            } else {
                Some(a.len() < b.len())
            }
        }
        (Signal::Num(a), b @ Signal::List(_)) => {
            compare_sigs(&Signal::List(vec![Signal::Num(*a)]), b)
        }
        (a @ Signal::List(_), Signal::Num(b)) => {
            compare_sigs(a, &Signal::List(vec![Signal::Num(*b)]))
        }
    }
}

#[derive(PartialEq, Eq)]
enum Token {
    Lparen,
    Rparen,
    Num(i32),
}

fn tokenize(mut b: Peekable<IntoIter<u8>>) -> BoxResult<Vec<Token>> {
    let mut tokens = Vec::new();
    while let Some(c) = b.next() {
        match c {
            b'[' => tokens.push(Token::Lparen),
            b']' => tokens.push(Token::Rparen),
            b'0'..=b'9' => {
                let mut buf = vec![c];
                while let Some(c2) = b.next_if(|&i| i >= b'0' && i <= b'9') {
                    buf.push(c2);
                }
                tokens.push(Token::Num(String::from_utf8(buf)?.parse::<i32>()?));
            }
            _ => (),
        }
    }
    Ok(tokens)
}

#[derive(Debug)]
enum Signal {
    Num(i32),
    List(Vec<Signal>),
}

fn parse_tokens(
    mut tokens: Peekable<IntoIter<Token>>,
) -> BoxResult<(Signal, Peekable<IntoIter<Token>>)> {
    let head = match tokens.next() {
        Some(v) => v,
        None => return Err("No tokens".into()),
    };
    match head {
        Token::Num(val) => Ok((Signal::Num(val), tokens)),
        Token::Lparen => {
            let mut signals = Vec::new();
            loop {
                match tokens.next_if(|t| t == &Token::Rparen) {
                    Some(_) => break,
                    None => (),
                };
                let (sig, rest) = parse_tokens(tokens)?;
                signals.push(sig);
                tokens = rest;
            }
            Ok((Signal::List(signals), tokens))
        }
        Token::Rparen => Err("Unexpected token".into()),
    }
}
